qiskit_experiments/framework/composite/composite_analysis.py

```diff
@@ -344,15 +344,37 @@ def _combine_results(
         """
         analysis_results = []
         figures = []
-        for i, sub_expdata in enumerate(component_experiment_data):
+        for sub_expdata in component_experiment_data:
             figures += sub_expdata._figures.values()
-            for result in sub_expdata.analysis_results():
-                # Add metadata to distinguish the component experiment
-                # the result was generated from
-                result.extra["component_experiment"] = {
-                    "experiment_type": sub_expdata.experiment_type,
-                    "component_index": i,
+
+            # Convert Dataframe Series back into AnalysisResultData
+            # This is due to limitation that _run_analysis must return List[AnalysisResultData],
+            # and some composite analysis such as TphiAnalysis overrides this method to
+            # return extra quantity computed from sub analysis results.
+            # This produces unnecessary data conversion.
+            # The _run_analysis mechanism seems just complicating the entire logic.
+            # Since it's impossible to deprecate the usage of this protected method,
+            # we should implement new CompositeAnalysis class with much more efficient
+            # internal logic. Note that the child data structure is no longer necessary
+            # because dataframe offers more efficient data filtering mechanisms.
+            analysis_table = sub_expdata.analysis_results(verbosity=3, dataframe=True)
+            for _, series in analysis_table.iterrows():
+                data_dict = series.to_dict()
+                primary_info = {
+                    "name": data_dict.pop("name"),
+                    "value": data_dict.pop("value"),
+                    "quality": data_dict.pop("quality"),
+                    "device_components": data_dict.pop("components"),
                 }
-                analysis_results.append(result)
+                chisq = data_dict.pop("chisq", np.nan)
+                if chisq:
+                    primary_info["chisq"] = chisq
+                data_dict["experiment"] = sub_expdata.experiment_type
+                if "experiment_id" in data_dict:
+                    # Use experiment ID of parent experiment data.
+                    # Sub experiment data is merged and discarded.
+                    del data_dict["experiment_id"]
+                analysis_result = AnalysisResultData(**primary_info, extra=data_dict)
+                analysis_results.append(analysis_result)
```

Comment:
I'm expecting this is a temporary conversion and will be removed after the breaking API change of the return type of `_run_analysis` planned in one of the following PRs. (I know the change is not easy...)

Reply:
These are removed with overhaul in [6ebd975](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1133/commits/6ebd975dfb19e4f6d4e43d2d100922ee9e1226ca).